File: modules/expo-shorts/ios/ExpoShortsModule.swift
import ExpoModulesCore
import AVFoundation

// -----------------------------------------------------------------------------
// ExpoShortsModule — Stage 2/3 (2026-05-16, 2026-05-18)
//
// Wires the native Shorts/Reels player on iOS:
//   - prefetchShortsVideo binds the URL to a pooled AVPlayer (paused).
//   - releasePool tears down all 3 AVPlayer instances.
//   - View prop setters route through AVPoolManager.
//
// Stage 3 additions:
//   - seek(viewTag, ms) — drives the JS scrubber.
//   - setAudioSessionPlayback / restoreAudioSession — TikTok parity, audio
//     keeps playing on lock screen / app background.
//   - View now emits `onTime` (~8fps) for the scrubber.
// -----------------------------------------------------------------------------

public class ExpoShortsModule: Module {
  // Remember the session state we found so restore() doesn't over-step into
  // settings the caller never picked.
  private var savedCategory: AVAudioSession.Category?
  private var savedMode: AVAudioSession.Mode?
  private var savedOptions: AVAudioSession.CategoryOptions = []
  private var didActivatePlayback: Bool = false

  public func definition() -> ModuleDefinition {
    Name("ExpoShorts")

    AsyncFunction("prefetchShortsVideo") { (id: String, url: String) in
      AVPoolManager.shared.prefetch(urlString: url)
      NSLog("[ExpoShorts] prefetchShortsVideo: id=\(id) url=\(url)")
    }

    AsyncFunction("releasePool") {
      AVPoolManager.shared.releaseAll()
      NSLog("[ExpoShorts] releasePool: drained")
    }

    // Switch AVAudioSession to .playback so Reels audio continues over the
    // lock screen and survives Spotify/podcast handoff the way TikTok does.
    // Idempotent: subsequent calls are a no-op.
    AsyncFunction("setAudioSessionPlayback") { [weak self] () -> Void in
      guard let self = self else { return }
      let s = AVAudioSession.sharedInstance()
      if !self.didActivatePlayback {
        self.savedCategory = s.category
        self.savedMode = s.mode
        self.savedOptions = s.categoryOptions
      }
      do {
        try s.setCategory(.playback, mode: .moviePlayback, options: [])
        try s.setActive(true, options: [])
        self.didActivatePlayback = true
        NSLog("[ExpoShorts] AVAudioSession → .playback")
      } catch {
        NSLog("[ExpoShorts] setAudioSessionPlayback error: \(error.localizedDescription)")
      }
    }

    // Roll the AVAudioSession back to whatever it was before Reels mounted.
    AsyncFunction("restoreAudioSession") { [weak self] () -> Void in
      guard let self = self else { return }
      guard self.didActivatePlayback else { return }
      let s = AVAudioSession.sharedInstance()
      do {
        if let cat = self.savedCategory {
          try s.setCategory(cat, mode: self.savedMode ?? .default, options: self.savedOptions)
        }
        try s.setActive(false, options: [.notifyOthersOnDeactivation])
      } catch {
        NSLog("[ExpoShorts] restoreAudioSession error: \(error.localizedDescription)")
      }
      self.didActivatePlayback = false
      self.savedCategory = nil
      self.savedMode = nil
      self.savedOptions = []
    }

    OnDestroy {
      AVPoolManager.shared.releaseAll()
    }

    View(ShortsPlayerView.self) {
      Events("onPlaybackReady", "onBuffering", "onError", "onTime")

      Prop("videoUrl") { (view: ShortsPlayerView, value: String?) in
        view.setVideoUrl(value)
      }

      Prop("playWhenInFocus") { (view: ShortsPlayerView, value: Bool?) in
        view.setPlayWhenInFocus(value ?? false)
      }

      Prop("muted") { (view: ShortsPlayerView, value: Bool?) in
        view.setMuted(value ?? true) // default muted (WhatsApp/IG)
      }

      Prop("playbackRate") { (view: ShortsPlayerView, value: Double?) in
        view.setPlaybackRate(value ?? 1.0)
      }

      // View-bound async function (Expo Modules dispatches to the view by its
      // RN tag). Drives the JS scrubber by seeking the bound AVPlayer.
      AsyncFunction("seek") { (view: ShortsPlayerView, ms: Double) in
        view.seek(toMillis: ms)
      }
    }
  }
}

File: modules/expo-shorts/android/src/main/java/expo/modules/shorts/ExpoShortsModule.kt
package expo.modules.shorts

import android.util.Log
import androidx.media3.common.util.UnstableApi
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

// -----------------------------------------------------------------------------
// ExpoShortsModule — Stage 2 (2026-05-16)
//
// Wires the native Shorts/Reels player. Stage 2 changes:
//   - prefetchShortsVideo now binds the URL to a free ExoPlayer in
//     ExoPoolManager and calls prepare() (no play).
//   - releasePool tears the whole pool down — call from JS on memory
//     warning / unmount of the Reels feed.
//   - View prop setters route through the pooled players.
// -----------------------------------------------------------------------------

@UnstableApi
class ExpoShortsModule : Module() {

  companion object {
    private const val TAG = "ExpoShorts"
  }

  override fun definition() = ModuleDefinition {
    Name("ExpoShorts")

    // Warm-up the player pool for `url`. Stage 1 was a no-op log.
    AsyncFunction("prefetchShortsVideo") { id: String, url: String ->
      try {
        val ctx = appContext.reactContext ?: return@AsyncFunction
        ExoPoolManager.prefetch(ctx, url)
        Log.d(TAG, "prefetchShortsVideo: id=$id url=$url")
      } catch (t: Throwable) {
        Log.w(TAG, "prefetchShortsVideo failed: ${t.message}")
      }
    }

    // Cleanup hook — call when the user navigates away from Reels.
    AsyncFunction("releasePool") {
      try {
        ExoPoolManager.release()
        Log.d(TAG, "releasePool: pool released")
      } catch (t: Throwable) {
        Log.w(TAG, "releasePool failed: ${t.message}")
      }
    }

    // Audio-session helpers — iOS sets AVAudioSession.playback so lock-screen
    // audio keeps going. Android handles that via the MediaSession + audio
    // focus in ExoPlayer, which is already on by default. We expose stubs so
    // the JS surface is platform-agnostic.
    AsyncFunction("setAudioSessionPlayback") {
      Log.d(TAG, "setAudioSessionPlayback: noop on Android (ExoPlayer handles focus)")
    }

    AsyncFunction("restoreAudioSession") {
      Log.d(TAG, "restoreAudioSession: noop on Android")
    }

    // Module lifecycle: drain the pool when the module is destroyed by RN
    // (full reload / app teardown).
    OnDestroy {
      try {
        ExoPoolManager.release()
      } catch (_: Throwable) {}
    }

    View(ShortsPlayerView::class) {
      Events("onPlaybackReady", "onBuffering", "onError", "onTime")

      Prop("videoUrl") { view: ShortsPlayerView, value: String? ->
        view.setVideoUrl(value)
      }

      Prop("playWhenInFocus") { view: ShortsPlayerView, value: Boolean? ->
        view.setPlayWhenInFocus(value == true)
      }

      Prop("muted") { view: ShortsPlayerView, value: Boolean? ->
        // Default true (WhatsApp/IG start muted)
        view.setMuted(value ?: true)
      }

      Prop("playbackRate") { view: ShortsPlayerView, value: Double? ->
        view.setPlaybackRate(value ?: 1.0)
      }

      // View-bound async function: drives the JS scrubber by seeking the
      // pooled ExoPlayer attached to this surface.
      AsyncFunction("seek") { view: ShortsPlayerView, ms: Double ->
        view.seekToMs(ms)
      }
    }
  }
}
